IPNS.md

```diff
@@ -29,95 +38,302 @@ Each time a file is modified, its content address changes. As a consequence, the
-## Protocol
+Notes:
 
-Taking into consideration a p2p network, each peer should be able to publish IPNS records to the network, as well as to resolve the IPNS records published by other peers.
+- For legacy reasons, some values must be stored in both `IpnsEntry` protobuf and `IpnsEntry.data` CBOR.
+  This should not be ignored, as it impacts interoperability with old software.
 
-When a node intends to publish a record to the network, an IPNS record needs to be created first. The node needs to have a previously generated asymmetric key pair to create the record according to the datastructure previously specified. It is important pointing out that the record needs to be uniquely identified in the network. As a result, the record identifier should be a hash of the public key used to sign the record.
+### Record Size Limit
 
-As an IPNS record may be updated during its lifetime, a versioning related logic is needed during the publish process. As a consequence, the record must be stored locally, in order to enable the publisher to understand which is the most recent record published. Accordingly, before creating the record, the node must verify if a previous version of the record exists, and update the sequence value for the new record being created.
+IPNS implementations MUST support sending and receiving a serialized
+`IpnsEntry` less than or equal to **10 KiB** in size.
 
-Once the record is created, it is ready to be spread through the network. This way, a peer can use whatever routing system it supports to make the record accessible to the remaining peers of the network.
+Records over the limit MAY be ignored. Handling records larger than the
+limit is not recommended so as to keep compatibility with implementations and
+transports that follow this specification.
 
-On the other side, each peer must be able to get a record published by another node. It only needs to have the unique identifier used to publish the record to the network. Taking into account the routing system being used, we may obtain a set of occurrences of the record from the network. In this case, records can be compared using the sequence number, in order to obtain the most recent one.
+### Backward Compatibility
 
-As soon as the node has the most recent record, the signature and the validity must be verified, in order to conclude that the record is still valid and not compromised.
+Implementations that want to interop with the public IPFS swarm MUST maintain
+backward compatibility for legacy consumers of IPNS records:
 
-Finally, the network nodes may also republish their records, so that the records in the network continue to be valid to the other nodes.
 
-## Overview
+- A legacy publisher MUST always be able to update to the latest implementation
+  of this specification without breaking record resolution for legacy consumers.
+- A legacy consumer MUST always be able to resolve IPNS name, even when publisher
+  updated to the latest implementation of this specification.
```

Comment:
## Some DHT  stats

We've looked at some stats from Hydra nodes (a small sample of recently valid records):
```
IPNS Records Signatures:
  v1:       9513
  v2:       5821
  with pk:  733 (RSA)
  nil data: 3692 (V1-only)
```

Every V2 record is also V1 (V1-only records have no data field, so their count is in  "nil data").
This means for the sample size of 9513 we have ~61% of V2+V1 and ~38% V1-only. 
We also looked at other time-ranges and V2 is growing, V1-only is decreasing – seems to follow the usual slow update curve we also see for Kubo itself.

## On this upgrade strategy

It is reasonable to expect both clients and  publishers will update over time, 
but the act of updating a publisher should not introduce any record resolution regressions for legacy nodes.

To make this clear, I've documented backward compatibility policy in https://github.com/ipfs/specs/pull/319/commits/b926e8c696a31686c26dfa931632cc307b411df5

> - A legacy publisher MUST always be able to update to the latest implementation
>   of this specification without breaking record resolution for legacy consumers.
> - A legacy consumer MUST always be able to resolve IPNS name, even when publisher
>   updated to the latest implementation of this specification.

This policy ensures the spec is aligned with existing code that was created to ensure IPNS implementations in GO and JS are backward-compatible from the perspective of legacy consumers:
- both Kubo and js-ipfs produce V1+V2 records
  - This allows legacy go-ipfs <0.9.0 node to correctly resolve IPNS records published by modern Kubo 0.15+
  - Maintaining this contract is important, I've added it to this spec
- Old go-ipfs <0.9.0 node publishing legacy RSA+V1-only records   should be able to upgrade to Kubo 0.16  and start publishing V1+V2 records, allowing for V2 resolution on modern clients (but not break V1-only for legacy ones).